dds/DCPS/XTypes/DynamicType.h

```diff
@@ -0,0 +1,49 @@
+#ifndef OPENDDS_DCPS_XTYPES_DYNAMIC_TYPE_H
+#define OPENDDS_DCPS_XTYPES_DYNAMIC_TYPE_H
+
+#include "External.h"
+#include "DynamicTypeMember.h"
+
+#include <dds/DCPS/RcHandle_T.h>
+#include <dds/DCPS/RcObject.h>
+#include <dds/DCPS/RcHandle_T.h>
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+namespace OpenDDS {
+namespace XTypes {
+
+class TypeDescriptor;
+
+class OpenDDS_Dcps_Export DynamicType : public OpenDDS::DCPS::RcObject
+{
+public:
+  DDS::ReturnCode_t get_descriptor(TypeDescriptor& descriptor);
+  OPENDDS_STRING get_name();
+  TypeKind get_kind();
+  DDS::ReturnCode_t get_member_by_name(DynamicTypeMember_rch& member, const OPENDDS_STRING& name);
+  DDS::ReturnCode_t get_all_members_by_name(DynamicTypeMembersByName& member);
+  DDS::ReturnCode_t get_member(DynamicTypeMember_rch& member, const MemberId& id);
+  DDS::ReturnCode_t get_all_members(DynamicTypeMembersById& member);
+  unsigned long get_member_count();
+  DDS::ReturnCode_t get_member_by_index(DynamicTypeMember_rch& member, unsigned long index);
+  bool equals(const DynamicType& other);
+  DynamicTypeMembersByName member_by_name;
+  DynamicTypeMembersById member_by_id;
+  DynamicTypeMembersByIndex member_by_index;
+  TypeDescriptor* descriptor_;
```

Comment:
We are ignoring annotation for our implementation according to our outline.

Making TypeDescriptor not a pointer requires the complete implementation rather than a declaration which causes circular include issues.

Reply:
Yeah, the design doc is ignoring `AnnotationDescriptor` but I think we need to implement it. It captures the annotations applied to types or members. If we ignore it, `DynamicType` will not have the complete information of the type, and equality checking can return an incorrect result.

Reply:
What annotations are you concerned about? 

We have tryconstruct, extensibility, nested, key, optional, must understand, shared, and default label as flags.

Reply:
Annotations like `@topic`, `@autoid`, `@hashid` haven't been captured. From 7.5.2.8.1, 7.5.2.8.4, 7.5.2.6.1, and 7.5.2.6.3 in the spec, I think all annotations applied to a type and its members need to be compared to determine equality. If we implement `AnnotationDescriptor` we have a general class that represents any annotation, so we don't have to enumerate any annotations currently supported or might be supported in the future. We can discuss further with the team.

Reply:
Use a smart pointer of some kind